File: dotnet/Visualizer/MainWindowController.cs
using System;
using Foundation;
using AppKit;
using CoreGraphics;
using Visualizer.ViewControllers;

namespace Visualizer
{
    [Register("MainWindowController")]
    public class MainWindowController : NSWindowController
    {
        private readonly ViewController _vc = new ViewController();
        private readonly ConnectViewController _connectViewController = new ConnectViewController();
        private readonly VisualizerViewController _visualizerViewcontroller = new VisualizerViewController();
        private readonly MetalTestViewController _metalTestViewController = new MetalTestViewController();

        public MainWindowController(IntPtr handle) : base(handle)
        {
        }

        [Export("initWithCoder:")]
        public MainWindowController(NSCoder coder) : base(coder)
        {
        }

        public MainWindowController() : base("MainWindow")
        {
            var contentRect = new CGRect(0, 0, 1000, 500);
            base.Window = new MainWindow(contentRect, NSWindowStyle.Titled | NSWindowStyle.Closable | NSWindowStyle.Miniaturizable | NSWindowStyle.Resizable, NSBackingStore.Buffered, false);

            Window.ContentView = _metalTestViewController.View;
            //Window.ContentView = _connectViewController.View;

            //Window.ContentView = _connectViewController.View;
            _connectViewController.OnConnect += OnConnect;
            _visualizerViewcontroller.OnDisconnect += OnDisconnect;
        }

        private void OnDisconnect(object sender, EventArgs e)
        {
            Window.ContentView = _connectViewController.View;
        }

        private void OnConnect(object sender, ConnectEventArgs e)
        {
            _visualizerViewcontroller.Connect(e.Host, e.Port);
            Window.ContentView = _visualizerViewcontroller.View;
        }

        public new MainWindow Window => (MainWindow)base.Window;
    }
}

File: apple/Visualizer/Visualizer/MainWindowController.swift
import Foundation
import CoreGraphics
import AppKit

class MainWindowController : NSWindowController {

    private let connectViewController: ConnectViewController
    private let visualizerViewController: VisualizerViewController
    required init(coder: NSCoder) {
        fatalError("No nibs here")
    }

    init(window: NSWindow) {
        connectViewController = ConnectViewController()
        visualizerViewController = VisualizerViewController()
        super.init(window: window)
        var frame = window.frame
        frame.size = NSSize(width: 1200, height: 1000)
        window.setFrame(frame, display: true)
        window.contentView = connectViewController.view
        connectViewController.delegate = self
        visualizerViewController.delegate = self
    }
}

extension MainWindowController : ConnectViewControllerDelegate {

    func connectViewController(_ controller: ConnectViewController, didConnectTo host: String, port: Int) {
        window?.contentView = visualizerViewController.view
    }
}

extension MainWindowController: VisualizerViewControllerDelegate {

    func visualizerViewController(_ controller: VisualizerViewController, disconnected: String) {
        print("alright!")
        window?.contentView = connectViewController.view
    }
}